### ISSUE TITLE
RedisTemplate.opsForValue().set(K key, V value) involves null value deserialization

### ISSUE THREAD

**Issue by xshipeng** [spring-projects/spring-data-redis#2511]
Hi, current opsForValue().set(K key, V value) overrides `inRedis` method and then passes `null` value to `deserializeValue` method. Is that intentional? Would it be better to overrdie `doInRedis` method in `RedisCallback` so that no deseralization is involved?
https://github.com/spring-projects/spring-data-redis/blob/f1f5f8da2568233fdd9a8372093c5908dfb10e35/src/main/java/org/springframework/data/redis/core/DefaultValueOperations.java#L240
I can help open a PR if needed.

**Comment by jxblum** [comment]
Hi @xshipeng -

Thank you for raising this concern.

I see what and where you are specifically referring to (using tagged version reference to code, so ref won't change), specifically [here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/core/DefaultValueOperations.java#L242), followed by [here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/core/AbstractOperations.java#L62).

However, if we follow the execution path through, then we observe [this](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/core/AbstractOperations.java#L356-L361), and if we specifically inspect different implementations of the `RedisSerializer` for "values" [ `ByteArrayRedisSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/ByteArrayRedisSerializer.java#L38-L39)), `GenericJackson2RedisSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/GenericJackson2JsonRedisSerializer.java#L214-L234)), `GenericToStringSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/GenericToStringSerializer.java#L72-L77)), `Jackson2JsonRedisSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/Jackson2JsonRedisSerializer.java#L129-L134)), `JdkSerializationRedisSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/JdkSerializationRedisSerializer.java#L75-L79)), `OxmSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/OxmSerializer.java#L87-L92)), `StringRedisSerializer` ([here](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/StringRedisSerializer.java#L83-L85)) ], then we notice that (currently) all of them are `null`-safe and even expect `@Nullable` (i.e. `null`) values, handling them appropriatetly.

Also, it's only a problem if a `RedisSerializer` for values [has been configured](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/core/AbstractOperations.java#L357-L359) in the first place.

So, if framework and user-provided implementations alike honor the `@Nullable` [contract](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/RedisSerializer.java#L41-L48) of the `deserialize(..)` method, then I think it is safe to say that `null` parameters and return values is a nil concern.

**Comment by xshipeng** [comment]
Hi @jxblum ,
Thank you for detailed response. 
I implemented the custom protostuff seralizer during development and got NullPointerException in unit testing for this `set` operation. After finding out the NPE was thrown in the custom `deserialize` function, I got confused since I assume that only `serialize` function will be called during set operation. 

By assuming users will always check for null value in custom `RedisSerializer` and follow @Nullable [contract](https://github.com/spring-projects/spring-data-redis/blob/3.0.2/src/main/java/org/springframework/data/redis/serializer/RedisSerializer.java#L41-L48), we should be good to close this issue.
